parlai/core/tod/tod_agents.py

```diff
@@ -0,0 +1,787 @@
+
+As a convention, agents and teachers that are inheritable are prefixed with "Tod" whereas those that can be used as-is are not. Similarly, classes and functions that do not need to be exposed outside of this file are prefixed with a single underscore ('_').
+
+## Why we do this
+These files aid in consistency between Teachers and Agents for simulation. Rather than having to align multiple different agents to be consistent about assuptions about data formatting, tokens, spacing, etc, we do this once (via converting everything to `TodStructuredEpisode`) and let the code handle the rest.
+
+# Description of Agents + Teachers useful for Simulation
+## Teachers for training (generative) models
+    * TodSystemTeacher
+    * TodUserSimulatorTeacher
+
+## Agents for Grounding
+For goal grounding for the User for simulation:
+    * TodGoalAgent
+    * TodSingleGoalAgent
+
+For (optional) API schema grounding for the System:
+    * TodApiSchemaAgent (must be used with `TodGoalAgent` only)
+    * TodSingleApiSchemaAgent (must be used with `TodSingleGoalAgent` only)
+    * EmptyApiSchemaAgent
+        * Used for simulations where the expectation is `no schema`, ie, evaluation simulations.
+
+## Agents for mocking APIs:
+    * StandaloneApiAgent
+         * Assumed to be provided a .pickle file 'trained' by `TodStandaloneApiTeacher`
+
+# Agents for dumping data from a ground truth dataset
+The following are for extracting TOD World metrics from a ground truth dataset. These are generally used sparingly and only for calculating baselines.
+    * TodApiCallAndSysUttAgent
+    * TodApiResponseAgent
+    * TodUserUttAgent
+
+For this metrics extraction, `TodGoalAgent` and `TodApiSchemaAgent` should be used.
+
+# Other agents
+There is a `EmptyGoalAgent` for use in human-human conversations where a goal is unnecessary.
+"""
+
+from parlai.core.agents import Agent
+from parlai.core.message import Message
+from parlai.core.metrics import AverageMetric
+from parlai.core.params import ParlaiParser
+from parlai.core.opt import Opt
+from parlai.core.teachers import DialogTeacher
+from parlai.utils.distributed import is_distributed, get_rank, num_workers
+
+import parlai.core.tod.tod_core as tod
+from parlai.core.tod.tod_core import SerializationHelpers
+from parlai.core.tod.teacher_metrics import SlotMetrics, NlgMetrics
+
+from typing import Optional, List
+import json
+import pickle
+import difflib
+import random
+from math import ceil
+
+
+######### Agents that dump information from a dataset; base classes
+class TodStructuredDataParser(Agent):
+    """
+    Base class that specifies intermediate representations for Tod conversations.
+    """
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        if hasattr(super(), "add_cmdline_args"):
```

Comment:
Hmm Agent doesn't have `self.add_cmdline_args()`, it looks like, but presumably this check is for multiple inheritance?

Reply:
Yeah, pretty much. Can add a comment to document this.